Additional MifareDesfireKey fixes + tests

darconeous opened the pull request:
In addition to adding tests for the bugs addressed by @mtdcr via #70, this commit also addresses a key corruption bug that would occur on 3DES keys when `mifare_desfire_key_set_version()` was called.

Merging this pull request should close #70.

darconeous commented:
I've run the cutter test suite with both a *MIFARE DESFire EV1 8k* card and a *MIFARE Ultralight C* to verify that the changes to the key behavior doesn't introduce any obvious regressions. Everything seems to work great.

smortex commented:
> this commit also addresses a key corruption bug that would occur on 3DES keys when mifare_desfire_key_set_version() was called.

Can you please elaborate on this? AFAICR, in order to distinguish DES keys from 3DES keys, the LSB of bytes [0..7] and [8..15] must be the same for DES and different for 3DES.

Apart from this points, this looks good to me :+1:

darconeous commented:
The key corruption bug corrupted much more than just the LSB of [0..7] and [8..15], and can be identified with the new tests. It was causing actual key bytes to be overwritten with 0xFF and 0xFE.

What code needs to look at those bits to differentiate between DES and 3DES? The `type` field seems like it does a good enough job differentiating the keys. If this is something that the DESFire card requires, then 1) I would think AN10922 would mention it along with the version hack, and 2) the test suite would fail. Neither of these are the case.

doegox commented:
Hi @darconeous, @smortex is referring to the following facts:

* DESFire uses the first 8 LSB to store the key version (one byte)
* Even if you don't use key versioning, you must overwrite the LSB (to e.g. 0) to avoi any information leak
* All other LSB are not used for key versioning
* **Even if the first 8 byte key itself is the same as the second 8 byte, but the version is coded
into one half, it is used as a 3DES key and not as a single DES key during authentication
and session key generation**

So if you want to use the key as DES key, you basically need to copy also the key versioning info into the LSB of the bytes 8..15.

darconeous commented:
> * DESFire uses the first 8 LSB to store the key version (one byte)

Check.

> * Even if you don't use key versioning, you must overwrite the LSB (to e.g. 0) to avoi any information leak

Check.

> * All other LSB are not used for key versioning

Check.

> * Even if the first 8 byte key itself is the same as the second 8 byte, but the version is coded
into one half, it is used as a 3DES key and not as a single DES key during authentication
and session key generation
>
> So if you want to use the key as DES key, you basically need to copy also the key versioning info into the LSB of the bytes 8..15.

I'm still not sure where the code is that depends on that behavior (as opposed to just looking at `key->type`), but I've updated the changes to ensure that 8..15 stay in sync with 0..7.

darconeous commented:
argh. one sec, will fix.

doegox commented:
I don't know if libfreefare has code depending on that behavior but the DESFire EV1 has for sure :)

darconeous commented:
> I don't know if libfreefare has code depending on that behavior but the DESFire EV1 has for sure 😊

I don't have access to the DESFire EV1 full data sheet, unfortunately. But I do (somehow) have access to the original DESFire data sheet. I would think for DES operations it would behave similarly to the EV1. I'll poke around in it to see if I can find any reference to the behavior you describe.

I'm curious what the actual behavioral difference would be. Theoretically it would just be a performance issue, since you are performing three DES operations instead of just one, but the behavior should be pretty identical aside from the key using more bytes. Is 3DES really that much slower than DES on DESFire EV1?

In any case, hopefully my latest change addresses your concerns.

doegox commented:
See https://scancode.ru/upload/iblock/de4/mifare_application_programming_guide_for_desfire_rev.e.pdf p14
(but in the sample the so-called single DES key doesn't have the same two halves :confused:)

darconeous commented:
I've reviewed that document and the source code and I feel that the current changes should address everyone's concerns.

darconeous commented:
oh wait I see what was the original problem. one sec and I'll have a better fix.

darconeous commented:
Ok, I think that one should be a winner.

doegox commented:
This makes much sense, thanks!
@smortex I let you do the final review & merge

smortex commented on libfreefare/mifare_desfire_key.c:
```diff
@@ -124,16 +127,21 @@ mifare_desfire_key_get_version(MifareDESFireKey key)
 	if (key->type == T_DES) {
 	    key->data[n + 8] = key->data[n];
-	} else {
+	} else if (key->type == T_3DES) {
```
If we start to have code `if (key->type == A) { ... } else if (key->type == B) { ... } else ...`, can you please turn this into a switch case without default like e.g. https://github.com/nfc-tools/libfreefare/blob/69bf99b33daf350d23aee8abffa96b4ed96151d1/libfreefare/mifare_desfire_crypto.c#L186-L195

That way, we have to have an explicit condition for each possible value, and if some new key type would be introduced in the future, the compiler would complain because all possible values are not handled in the switch statement.

darconeous replied:
No problem, but how did you quote that other file inline in your comment like that?

darconeous replied:
In this case, we really only care about DES keys, and hopefully there are no fancy new DES encryption types that will be introduced. I think that is a safe assumption. Forcing a warning for every newly added encryption type doesn't seem necessary in this block.

smortex replied:
I was wondering what was happening with 3K3DES… Correct me if I am wrong, but these manipulation only have the role of making a DES key stay a DES key and a 3DES key stay a 3DES key when the key version is changed; a problem the 3K3DES keys can not be affected by.

If it's the reason 3K3DES are not handled here, maybe adding a comment here may be enough: AES keys are handled at the top of the function and having to list them again seems tedious, I quite agree; and as you say, support for more types of keys seems improbable.

[oh, and to include a snippet in a comment, I just write the URL of a file with the relevant lines highlighted (the snippets title links to the URL I typed :stuck_out_tongue_winking_eye: )]

smortex commented on libfreefare/mifare_desfire_key.c:
```diff
@@ -124,16 +127,21 @@ mifare_desfire_key_get_version(MifareDESFireKey key)
 	    // Write ~version to avoid turning a 3DES key into a DES key
 	    key->data[n + 8] &= 0xfe;
-	    key->data[n + 8] |= ~version_bit;
+	    key->data[n + 8] |= !version_bit;
```
D'oh :scream: !

So, yes, `~0x00` → `0xFF` and `~0x01` → `0xFE` so all key data is corrupt, this is so obvious I failed to see it…

That being said, or'ing with the result of a boolean expression looks odd to me: I am not 100% sure that C states that the result of a boolean expression is always 0 or 1, and if it's not guaranteed, it would be "safer" to:

```c
key->data[n + 8] |= (version_bit) ? 0x00 : 0x01;
```

I failed to find something that tells me that any boolean expression shall return 0 or 1.  If you have some information in respect to this, please educate me and ignore the preceding :smile:

darconeous replied:
IEEE Std 1003.1-2001 defines a boolean `true` to be 1 and `false` to be zero. So, yes, I'm 100% sure this isn't relying on undefined behavior.

smortex replied:
Great!  I have to admit that I feel like the "long" version is easier to read than the '!' one, but I can live with it :+1:

smortex commented on test/test_mifare_desfire_key.c:
```diff
@@ -65,7 +74,7 @@ test_mifare_desfire_key(void)
     mifare_desfire_key_free(key);
 
-    uint8_t key3_3des_data[16] = { 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0X00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x77 };
+    uint8_t key3_3des_data[16] = { 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x77 };
```
:+1: for minor style fix (there is at least another one at the same position on line 65 :wink:

darconeous replied:
fixed.

smortex commented:
> I don't have access to the DESFire EV1 full data sheet, unfortunately.

I guess this is not exactly what you are looking for, but most of the DESFire code was written using this: [MF3ICD81 - MIFARE DESFire - Functional specification - Rev. 3.5 - 28 November 2008](https://marvin.blogreen.org/~romain/nfc/MF3ICD81%20-%20MIFARE%20DESFire%20-%20Functional%20specification%20-%20Rev.%203.5%20-%2028%20November%202008.pdf)

darconeous commented:
I think the assumption regarding why the parity bits on bytes [8..23] of 3K3DES keys are not manipulated is at least corroborated by the fact that the 3K3DES key setter just zeros out the parity bits of [0..7] without touching the rest of the key (which is unlike the DES and 2K3DES setters, which each do something):

https://github.com/darconeous/libfreefare/blob/5908f367c15d0a8143d41be2fab3034e47cc92e5/libfreefare/mifare_desfire_key.c#L76

smortex commented on libfreefare/mifare_desfire_key.c:
```diff
@@ -124,16 +127,26 @@ mifare_desfire_key_get_version(MifareDESFireKey key)
-	} else {
+	    break;
+	case T_3DES:
 	    // Write ~version to avoid turning a 3DES key into a DES key
```
"Now that the parity bits of key bites [0..7] are set to version, write ~version in parity bits of key bytes [8..15] in order to avoid 3DES keys to become DES keys at the PICC level."

darconeous replied:
I've clarified the comment.

smortex commented:
Merged, thanks! :heart: :yellow_heart: :green_heart: :blue_heart:  :purple_heart:
